service/evaluation/evaluation.go

```diff
@@ -0,0 +1,845 @@
+	if req.GetInterval() == 0 {
+		interval = defaultInterval
+	} else {
+		interval = int(req.GetInterval())
+	}
+
+	// Get orchestrator client. The orchestrator client is used to retrieve necessary information from the Orchestrator, such as assessment_results, controls or targets_of_evaluation.
+	err = s.initOrchestratorClient()
+	if err != nil {
+		err = fmt.Errorf("could not set orchestrator client: %v", err)
+		log.Error(err)
+		return nil, status.Errorf(codes.Internal, "%s", err)
+	}
+
+	// Get Target of Evaluation. The Target of Evaluation is retrieved to get the controls_in_scope, which are then evaluated.
+	toe, err = s.orchestratorClient.GetTargetOfEvaluation(context.Background(), &orchestrator.GetTargetOfEvaluationRequest{
+		CloudServiceId: req.GetCloudServiceId(),
+		CatalogId:      req.GetCatalogId(),
+	})
+	if err != nil {
+		err = fmt.Errorf("could not get target of evaluation: %v", err)
+		log.Error(err)
+		return nil, status.Errorf(codes.Internal, "%s", err)
+	}
+
+	log.Info("Starting evaluation ...")
+
+	// Scheduler tags must be unique to find the jobs by tag name
+	s.scheduler.TagsUnique()
+
+	// Add the controls_in_scope of the target_of_evaluation including their sub-controls to the scheduler. The parent control has to wait for the evaluation of the sub-controls. That's why we need to know how much sub-controls are available and define the waitGroup with the number of the corresponding sub-controls. The controls_in_scope are not stored in a hierarchy, so we have to get the parent control and find all related sub-controls.
+	controlsInScope := getControlsInScopeHierarchy(toe.GetControlsInScope())
+	for _, control := range controlsInScope {
+		var (
+			controls []*orchestrator.Control
+		)
+
+		// If the control does not have sub-controls don't start a scheduler.
+		if control.ParentControlId != nil {
+			continue
+		}
+
+		// Collect control including sub-controls from the controls_in_scope slice to start them later as scheduler jobs
+		controls = append(controls, control)
+		controls = append(controls, control.GetControls()...)
+
+		// Check if scheduler job for current control_id is already running
+		jobs, err := s.scheduler.FindJobsByTag(schedulerTag)
+		if len(jobs) > 0 && err == nil {
+			err = fmt.Errorf("evaluation for Cloud Service '%s' and Control ID '%s' already started", toe.GetCloudServiceId(), control.GetId())
+			log.Error(err)
+			return nil, status.Errorf(codes.AlreadyExists, "%s", err)
+		}
+
+		// parentSchedulerTag is the tag for the parent control (e.g., OPS-13)
+		parentSchedulerTag = getSchedulerTag(toe.GetCloudServiceId(), control.GetId())
+
+		// Add number of sub-controls to the WaitGroup. For the control (e.g., OPS-13) we have to wait until all the sub-controls (e.g., OPS-13.1) are ready.
+		// Control is a parent control if no parentControlId exists.
+		s.wg[parentSchedulerTag] = &WaitGroup{
+			wg:      &sync.WaitGroup{},
+			wgMutex: sync.Mutex{},
```

Comment:
What's the point of the mutex *inside* the map? Shouldn't the mutex protect the map? The wait group itself should already be thread-safe. Not quite sure what you want to achieve here :)

Reply:
Thatś right, I think this is an old mutex from the beginning that is now obsolete.  I will change that and try :)